examples/ecs/computed_states.rs

```diff
@@ -0,0 +1,675 @@
+// needing to compare against every version of the AppState that could involve them.
+//
+// In addition, it allows us to still maintain a strict type representation - you can't Turbo
+// if you aren't in game, for example - while still having the
+// flexibility to check for the states as if they were completely unrelated.
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+struct TurboMode;
+
+impl ComputedStates for TurboMode {
+    type SourceStates = AppState;
+
+    fn compute(
+        sources: <<Self as ComputedStates>::SourceStates as bevy_internal::ecs::schedule::StateSet>::Optionals,
+    ) -> Option<Self> {
+        match sources {
+            Some(AppState::InGame { turbo: true, .. }) => Some(Self),
+            _ => None,
+        }
+    }
+}
+
+// For the [`IsPaused`] state, we'll actually use an `enum` - because the difference between `Paused` and `NotPaused`
+// involve activating different systems.
+//
+// To clarify the difference, `InGame` and `TurboMode` both activate systems if they exist, and there is
+// no variation within them. So we defined them as Zero-Sized Structs.
+//
+// In contrast, pausing actually involve 3 distinct potential situations:
+// - it doesn't exist - this is when being paused is meaningless, like in the menu.
+// - it is `NotPaused` - in which elements like the movement system are active.
+// - it is `Paused` - in which those game systems are inactive, and a pause screen is shown.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+enum IsPaused {
+    NotPaused,
+    Paused,
+}
+
+impl ComputedStates for IsPaused {
+    type SourceStates = AppState;
+
+    fn compute(
+        sources: <<Self as ComputedStates>::SourceStates as bevy_internal::ecs::schedule::StateSet>::Optionals,
+    ) -> Option<Self> {
+        // Here we convert from our [`AppState`] to all potential [`IsPaused`] versions.
+        match sources {
+            Some(AppState::InGame { paused: true, .. }) => Some(Self::Paused),
+            Some(AppState::InGame { paused: false, .. }) => Some(Self::NotPaused),
+            // If `AppState` is not `InGame`, pausing is meaningless, and so we set it to `None`.
+            _ => None,
+        }
+    }
+}
+
+// Lastly, we have our tutorial, which actually has a more complex derivation.
+//
+// Like `IsPaused`, the tutorial has a few fully distinct possible states, so we want to represent them
+// as an Enum. However - in this case they are all dependant on multiple states: the root [`TutorialState`],
+// and both [`InGame`] and [`IsPaused`] - which are in turn derived from [`AppState`].
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+enum Tutorial {
+    MovementInstructions,
+    PauseInstructions,
+}
+
+impl ComputedStates for Tutorial {
+    // We can also use tuples of types that implement [`States`] as our [`SourceStates`].
+    // That includes other [`ComputedStates`], but be careful of circular dependencies.
```

Comment:
Is there anything to check for circular dependencies?
Worth noting down

Reply:
At the time we couldn't, but circular dependencies won't compile!